### DISCUSSION TITLE
Request: move the Serial/UART related code out of the Terminal class

### DISCUSSION THREAD

**Discussion by dmcnaugh** [fdivitto/FabGL#164] [Ideas]
It would be useful if all the Serial/UART related code currently in `class Terminal` was moved out to a new derived class e.g. `class SerialTerminal : public Terminal`

That would mean moving:

```
HardwareSerial *   m_serialPort;
void connectSerialPort(Stream & serialPort, bool autoXONXOFF = true);
void pollSerialPort();
```
and the new UART2 related functions
along with the current versions of:
```
  void send(char c);
  void send(char const * str);
```

into this new class.

The 2 `send()` functions would then need to be made `protected` and `virtual` in `class Terminal` so that they can be polymorphic. Alternatively the two `localwrite()` functions could be renamed as `send()` and simply overridden if required.

These thoughts may not be perfect, but the end goal is that the `AnsiTerminal` and the `NetworkTerminal` examples should deal with the `Terminal` class the same way, whereas at the moment a serial connected `Terminal` is a special case of a `Terminal` instance with the code that handles the serial connection I/O inside the class, but a network connected Terminal manages the network connection I/O itself. A serial connected `Terminal`should not need to be special.

**Comment by fdivitto** [comment]
Serial/UART will be moved out of Terminal class, I'm working on...

**Comment by fdivitto** [comment]
...just implemented.
Now to connect a serial port to Terminal object, an intermediate object (SerialPortTerminalConnector) is required. This will make design more clear (and versatile) at the cost of a few more lines of code.
For example, declare following objects:

```
fabgl::Terminal                     Terminal;
fabgl::SerialPort                   SerialPort;
fabgl::SerialPortTerminalConnector  SerialPortTerminalConnector;
```

Connect serialport to the terminal and set communication parameters:

```
SerialPortTerminalConnector.connect(&SerialPort, &Terminal);
SerialPort.setup(2, 115200, 8, 'N', 1, 34, 2, FlowControl::Software);
```

connect() method must be called after Terminal.begin(), like the updated AnsiTerminal demo does.
